src/ucs/datastruct/string_buffer.c

```diff
@@ -17,116 +17,89 @@
 
+/* Minimal reserve size when appending new data */
+#define UCS_STRING_BUFFER_RESERVE  32
 
-static void ucs_string_buffer_reset(ucs_string_buffer_t *strb)
-{
-    strb->buffer   = NULL;
-    strb->length   = 0;
-    strb->capacity = 0;
-}
+UCS_ARRAY_FUNCS(string_buffer, size_t, char, static UCS_F_ALWAYS_INLINE)
 
 void ucs_string_buffer_init(ucs_string_buffer_t *strb)
 {
-    ucs_string_buffer_reset(strb);
+    ucs_array_init_dynamic(string_buffer, &strb->str);
 }
 
 void ucs_string_buffer_cleanup(ucs_string_buffer_t *strb)
 {
-    ucs_free(strb->buffer);
-    ucs_string_buffer_reset(strb);
+    ucs_array_cleanup_dynamic(string_buffer, &strb->str);
 }
 
-/*
- * Grow the buffer to at least the required size and at least double the
- * previous size (to reduce the amortized cost of realloc)
- */
-static ucs_status_t ucs_string_buffer_grow(ucs_string_buffer_t *strb,
-                                           size_t min_capacity)
+size_t ucs_string_buffer_length(ucs_string_buffer_t *strb)
 {
-    size_t new_capacity;
-    char *new_buffer;
-
-    new_capacity = ucs_max(strb->capacity * 2, min_capacity);
-    new_buffer   = ucs_realloc(strb->buffer, new_capacity,
-                               UCS_STRING_BUFFER_ALLOC_NAME);
-    if (new_buffer == NULL) {
-        ucs_error("failed to grow string from %zu to %zu characters",
-                  strb->capacity, new_capacity);
-        return UCS_ERR_NO_MEMORY;
-    }
-
-    strb->buffer   = new_buffer;
-    strb->capacity = new_capacity;
-    /* length stays the same */
-    return UCS_OK;
+    return ucs_array_length(&strb->str);
 }
 
-ucs_status_t ucs_string_buffer_appendf(ucs_string_buffer_t *strb,
-                                       const char *fmt, ...)
+void ucs_string_buffer_appendf(ucs_string_buffer_t *strb, const char *fmt, ...)
 {
     ucs_status_t status;
     size_t max_print;
     va_list ap;
     int ret;
 
-    /* set minimal initial size */
-    if ((strb->capacity - strb->length) <= 1) {
-        status = ucs_string_buffer_grow(strb,
-                                        strb->capacity + UCS_STRING_BUFFER_GROW);
-        if (status != UCS_OK) {
-            return status;
-        }
-    }
+    ucs_array_reserve(string_buffer, &strb->str,
+                      ucs_array_length(&strb->str) + UCS_STRING_BUFFER_RESERVE);
 
     /* try to write to existing buffer */
     va_start(ap, fmt);
-    max_print = strb->capacity - strb->length - 1;
-    ret       = vsnprintf(strb->buffer + strb->length, max_print, fmt, ap);
+    max_print = ucs_array_end_capacity(&strb->str);
+    ret       = vsnprintf(ucs_array_end(&strb->str), max_print, fmt, ap);
     va_end(ap);
 
     /* if failed, grow the buffer accommodate for the expected extra length */
     if (ret >= max_print) {
-        status = ucs_string_buffer_grow(strb, strb->length + ret + 1);
+        status = ucs_array_reserve(string_buffer, &strb->str,
+                                   ucs_array_length(&strb->str) + ret + 1);
```

Comment:
minor: after this operation actual string buffer size will be ```size(original_array) + UCS_STRING_BUFFER_RESERVE + ret``` - we will have UCS_STRING_BUFFER_RESERVE non-used bytes

Reply:
why? we didn't change the length during the 1st ucs_array_reserve(), it only increased the capacity
the 2nd time ucs_array_reserve() is used, ucs_array_length() will still be the original length